packages/patrol_cli/lib/src/coverage/vm_connection_details.dart

```diff
@@ -0,0 +1,36 @@
+class VMConnectionDetails {
+  const VMConnectionDetails({
+    required this.port,
+    required this.auth,
+  });
+
+  final int port;
+  final String auth;
+
+  Uri get uri => Uri.parse('http://127.0.0.1:$port/$auth');
+  Uri get webSocketUri {
+    final pathSegments = uri.pathSegments.where((c) => c.isNotEmpty).toList()
+      ..add('ws');
+    return uri.replace(scheme: 'ws', pathSegments: pathSegments);
+  }
+
+  static VMConnectionDetails? tryExtractFromLogs(String logsLine) {
```

Comment:
A few tests for this would be nice to see

Reply:
I refactored the method to make it use `Uri.parse` instead of parsing manually. I added two very simple test cases. Any specific test cases you had in mind?

Reply:
Nope, the ones you added look good enough!